### PR TITLE
Add paged / block-table KV cache export (--features paged-cache)

### PR DESCRIPTION
**Pull request by justinchuby**
## What this adds

A **paged / block-table KV cache** export mode for causal LMs — a variant of the
existing `--static-cache` that emits models whose attention reads KV from
**non-contiguous pages** via a page pool + `block_table` + `slot_mapping`.

This is the **vLLM PagedAttention** layout, and because sequences can list the
*same* physical page in their `block_table`, the identical graph also expresses
**SGLang RadixAttention** (shared-prefix pages) with no change. It implements
onnx-genai `docs/DESIGN.md` §39.4 **Option C** ("ONNX Scatter/GatherElements in
Graph") using **only standard ONNX ops** (no custom op).

## Why

The onnx-genai runtime needs models that operate directly on a paged KV pool so
it can do vLLM-style paged attention and SGLang-style prefix sharing without
gathering pages into a contiguous buffer in the runtime on every step.

## How it works (per attention layer)

```
1. ScatterND(pool_flat, slot_mapping, new_kv)      # write new tokens to physical slots
2. Gather(updated_pool, block_table, axis=0)       # assemble this sequence's pages…
   → Reshape [1, num_blocks*page_size, kv_hidden]  # …into a contiguous KV sequence
3. Attention(query, K_gathered, V_gathered,
             nonpad_kv_seqlen, is_causal=1)         # opset-24 external-cache Attention
```

Step 3 is the **same op contract as `--static-cache`** (`Attention` with
`is_causal=1` + `nonpad_kv_seqlen` input #6, no `attention_mask`), so it reuses
the existing opset-24 retention logic (`_graph_requires_opset24` already detects
`Attention` with a non-empty input #6) and the same DecoderLayer/MoEDecoderLayer
dispatch pattern as the static cache.

Ops used: `Reshape`, `Shape`, `Unsqueeze`, `ScatterND` (write), `Gather`
(page assemble), `Attention`. All opset-24 compatible.

## New flag

```
mobius build <model> --paged-cache [--page-size N] [--num-pages N]
```
Also available programmatically: `CausalLMTask(paged_cache=True, page_size=16, num_pages=None)`.
Mutually exclusive with `--static-cache`; requires `DecoderLayer` / `MoEDecoderLayer` models.

## ONNX I/O contract (what onnx-genai must drive)

Single active sequence (`batch == 1`), per layer `i`, `kv_hidden = num_kv_heads * head_dim`:

**Inputs**
| name | shape | dtype |
|---|---|---|
| `input_ids` | `[batch, seq_len]` | int64 |
| `position_ids` | `[batch, seq_len]` | int64 |
| `key_pool.{i}` / `value_pool.{i}` | `[num_pages, page_size, kv_hidden]` | model dtype |
| `block_table` | `[num_blocks]` | int64 (physical page ids, logical order) |
| `slot_mapping` | `[seq_len]` | int64 (`page_id*page_size + offset` per new token) |
| `nonpad_kv_seqlen` | `[batch]` | int64 (`write_start + valid_token_count`) |

**Outputs**
| name | shape |
|---|---|
| `logits` | `[batch, seq_len, vocab]` |
| `updated_key_pool.{i}` / `updated_value_pool.{i}` | `[num_pages, page_size, kv_hidden]` |

No `attention_mask` — causal + padding masking is derived from `is_causal=1` +
`nonpad_kv_seqlen`. RoPE is baked into the keys before they are written to the
pool. RadixAttention: point multiple sequences' `block_table` at the same
physical page — the graph gathers it identically for each.

## Complete vs TODO

**Complete**
- Paged attention body (ScatterND write + Gather assemble + Attention) in `components/_attention.py`
- `PagedCacheState` dispatch in `DecoderLayer` and `MoEDecoderLayer` (backward-compatible: `paged_cache` only forwarded to custom `self_attn` modules when set)
- `CausalLMTask(paged_cache=…)` + input/output wiring, mutual-exclusion with static cache
- CLI flags `--paged-cache` / `--page-size` / `--num-pages` with validation
- Graph builds, `onnx.checker` passes, shape inference yields correct pool shapes
- Standard-Attention (dense) and MoE model paths

**TODO** (documented in code)
- Multi-sequence batching (2-D `block_table` / per-row gather); current impl targets `batch == 1`
- End-to-end paged **execution** parity: the `Attention` op with `nonpad_kv_seqlen` is CUDA-only until onnxruntime#28958 ships (same constraint as `--static-cache`), so full-model run parity is deferred to the shared static-cache Flash probe. The paging ops themselves are validated on CPU.

## Validation

- `tests/build_graph_test.py::TestBuildPagedCacheGraph` — 12 graph-level tests
  (inputs/outputs, pool shapes, dynamic `num_pages`, `ScatterND`/`Gather`/`Attention`
  presence, `is_causal=1`, `nonpad_kv_seqlen` wiring, MoE, mutual exclusion, shape inference).
- `tests/paged_cache_test.py` — CPU numerical parity for the scatter+gather paging
  ops vs a NumPy reference, including a **RadixAttention shared-page** case, plus an
  end-to-end `onnx.checker` + shape-inference build check.
- Full `TestBuildGraph` / `TestBuildStaticCacheGraph` / `cli_test` suites pass (no regressions).

> Draft: single-sequence paged cache is complete and validated; multi-sequence batching remains a TODO.

### CODE REVIEW

**Comment by justinchuby** [comment]
Not chasing in favor of the VMM solution.

**Comment by justinchuby** [comment]
Updated from `main` and folded the paged KV-cache toggle into the cargo-style `--features` option (added on main):

- `mobius build --features paged-cache` now enables the paged / block-table KV cache, consistent with `static-cache` / `fp8-kv-cache` / `prune-lm-head` / `text-only`.
- Removed the standalone `--paged-cache` boolean flag (main dropped the other boolean feature flags); `--page-size` / `--num-pages` remain as tuning parameters.
- Reworded validation errors + docs (README, cli_reference, CHANGELOG) to reference `--features paged-cache`; added a CLI test.
- Resolved the two prior review threads (attention_bias rejection in paged mode; `_validate_static_cache_support` mode-aware error text) — both already implemented on the branch.

Merge resolved conflicts in `__main__.py`, `tasks/_causal_lm.py`, and `CHANGELOG.md`; branch is now even with `main`. Tests: cli/paged/build-graph green.
